### PR TITLE
Feature/in memory cache for multio simulations

### PR DESCRIPTION
**Pull request by MircoValentiniECMWF**
Still rude implementation. Some notes:

1) An std::map has been chosen over an std::unordered_map.  The reason is that this map is supposed to contain never more than 5 elements since the grid resolution remain constant during a simulation. In scenarios with a small number of elements like this, the computational cost associated with the hash algorithm in an unordered map can sometimes exceed the efficiency gains achieved through key-comparisons.  Open to discuss this point.

2) The CMakeFiles.txt may seem unwieldy with its extensive if-else structure. Unfortunately, finding a more elegant solution proved challenging. Open to suggestions for improvement in this area.

3) Although using pointers within the map could potentially improve performance by reducing the need for copies, implementing this change would require more extensive code modifications. For the sake of maintaining code stability, I decided against such invasive changes. Open to suggestions for improvement in this area.

I tested on my personal machine and this modification gives a 10% to 20% boost on pure interpolation timing.
We are still working to perform a benchmark with the full coupled simulation.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/atlas-orca?pullRequest=14) <br/>All committers have signed the CLA.

File: CMakeLists.txt
Comment (dsarmany) [inline CMakeLists.txt:29]: Is there a risk of always using the cache? Even if not, a runtime option, potentially through configuration may make more sense. Any views, @wdeconinck ?

File: src/atlas-orca/grid/Orca.cc
Comment (dsarmany) [inline src/atlas-orca/grid/Orca.cc:40]: I agree `std::map` is the right choice here.

(But, as above, I'm not sure I like the conditional compilation.)

**Review by dsarmany — COMMENTED** [review]
I think I'd prefer a runtime flag, either through configuration or an environment variable. But I'll leave this to @wdeconinck to decide.

File: CMakeLists.txt
Comment (wdeconinck) [reply CMakeLists.txt:29]: I definitely think this should be a runtime option as well.

File: src/atlas-orca/CMakeLists.txt
Comment (wdeconinck) [inline src/atlas-orca/CMakeLists.txt:92]: I would avoid the code duplication.
Anyhow, seeing that a runtime option is preferred, this is not needed.
For future reference, you can just use a variable which can be empty:
```
ecbuild_add_library( ...
   PRIVATE_DEFINITIONS ${ATLAS_ORCA_PRIVATE_DEFINITIONS})
```
The definition should not be PUBLIC, but PRIVATE in this case.
You can also add per source file definitions:
```
if ( HAVE_IN_MEMORY_CACHE )
    set_source_files_properties( grid/Orca.cc
                            PROPERTIES COMPILE_DEFINITIONS "ATLAS_ORCA_HAVE_IN_MEMORY_CACHE")
endif()

**Review by wdeconinck — CHANGES_REQUESTED** [review]
Thanks Mirco for this.
See individual comments.
What's missing is a means to clear the cache.
We should be able to clear it when wanted, and it should be cleared upon plugin library destruction.

File: CMakeLists.txt
Comment (wdeconinck) [reply CMakeLists.txt:29]: Also there may be some issues in multithreaded context, on insertion in the map.

File: src/atlas-orca/grid/Orca.cc
Comment (MircoValentiniECMWF) [reply src/atlas-orca/grid/Orca.cc:40]: I believe that passing a variable through all the constructors may not be a feasible approach. Would it be more practical to consider using an environment variable instead?

File: src/atlas-orca/CMakeLists.txt
Comment (MircoValentiniECMWF) [reply src/atlas-orca/CMakeLists.txt:92]: I completely agree that code duplication is undesirable. I wasn't aware that I could have placed the definition in a specific source file.

Comment (pmaciel) [reply src/atlas-orca/CMakeLists.txt:92]: ~~You can look at mir's src/mir/caching/SharedMemoryKey.cc and src/mir/CMakeLists.txt, it's this exact situation too.~~
Actually I retract the comment. I completely agree with @wdeconinck that this functionality if we chose to add it should be a run-time option.

File: CMakeLists.txt
Comment (pmaciel) [reply CMakeLists.txt:29]: Indeed, there are multi-thhreading issues with the map. For consistency (with the rest of eckit-based software stack) you shold protect this with eckit::Mutex and not the std:: version, until that is fixed upstream.

**Comment by pmaciel** [comment]
1. The std::map choice is the good one, the advantages of std::unordered_map are marginal here.
2. CMakeFiles.txt is unwieldy yes -- suggestion above
3. As with 1., let's optimise where it's costly

However, I'm not so convinced I understand the PR itself, to add in-memory caching for reading ORCA grids? I see this plugin as a black-box for dependencies, as in, client code shouldn't directly concern with Atlas's plugins directly. This cache mapping could be downstream at the client code level, where one could cache any grid. Even at Atlas's level this would make more sense (as an optional, runtime mechanism in a factory.)

I think a broader discussion is needed. There are grids other than ORCA that are heavy to load, and a performance measurement should not include interpolation -- it's an orthogonal concern. Sure you see a benefit there, but that's translated into the overall run the same way. So, fine, if this solves a current problem, but I think there are more impactful ways to solve it.

**Comment by dsarmany** [comment]
> However, I'm not so convinced I understand the PR itself, to add in-memory caching for reading ORCA grids? I see this plugin as a black-box for dependencies, as in, client code shouldn't directly concern with Atlas's plugins directly. This cache mapping could be downstream at the client code level, where one could cache any grid. Even at Atlas's level this would make more sense (as an optional, runtime mechanism in a factory.)

I agree that at Atlas's level this should be a runtime option, ideally in the factory. Whether the client code itself should implement the caching instead is debatable. In any case, the direct client here is mir and not multio. This is all in the context of interpolation/re-gridding and not that of fetching the co-ordinates to encode in grib2 and output it to disk/FDB. So caching in MIR perhaps is a better solution.

> I think a broader discussion is needed. There are grids other than ORCA that are heavy to load, and a performance measurement should not include interpolation -- it's an orthogonal concern. Sure you see a benefit there, but that's translated into the overall run the same way. So, fine, if this solves a current problem, but I think there are more impactful ways to solve it.

See above. My understanding is that this is very much in the context of interpolation and the behaviour of mir re-loading the grids each time we carry it out. Yes, it is because we set up a separate MIRJob for each interpolation. Is it something we shouldn't do? Perhaps, but I believe it was discussed before and the upshot was that otherwise we'd need to cache a MIRJob for all different types of source grids, etc. A local cache closer to where the grid is actually loaded/created (in the factory indeed) seems sensible. Happy to discuss.

**Comment by MircoValentiniECMWF** [comment]
@wdeconinck It is evident that we require a mechanism to clear the cache. However, due to the current approach of invoking mir from multIO, it cannot be integrated into the ORCA destructor. In my view, the most appropriate time to clear the cache would be during the plugin deregistration process. This perspective aligns with the idea that caching should be a runtime option within the Atlas Grid factory, coupled with a logic to determine which grids should be cached (i.e. grids that require I/O)."

**Comment by tlmquintino** [comment]
Please make this a runtime option. 
Code in CMake like `ORCA_IN_MEMORY_CACHE` going down to `#if defined(HAVE_ORCA_IN_MEMORY_CACHE)` is not desirable.

File: src/atlas-orca/CMakeLists.txt
Comment (tlmquintino) [reply src/atlas-orca/CMakeLists.txt:92]: Agree. See my comment below.

**Comment by tlmquintino** [comment]
> Thanks Mirco for this.
> See individual comments.
> What's missing is a means to clear the cache.
> We should be able to clear it when wanted, and it should be cleared upon plugin library destruction.

Agreed. we need to be able to control caches including choosing to evict (clear).

**Comment by wdeconinck** [comment]
I also tend to agree with @pmaciel  in fact that caching would better be done at a higher level, e.g. the Grid object, and the same could be said about e.g. an expensive Mesh generation step. That would trigger a whole design change in multiple packages.
e.g. MIR keeping an in-memory cache of grid objects alive, across invocations. This would then indeed also cater for any unstructured grid.
From a design point it is rather obscure to have to rely on plugin or specialised grid construction mechanisms to do caching itself for performance.

**Comment by tlmquintino** [comment]
> I also tend to agree with @pmaciel in fact that caching would better be done at a higher level, e.g. the Grid object, and the same could be said about e.g. an expensive Mesh generation step. That would trigger a whole design change in multiple packages.
> e.g. MIR keeping an in-memory cache of grid objects alive, across invocations. This would then indeed also cater for any unstructured grid.
> From a design point it is rather obscure to have to rely on plugin or specialised grid construction mechanisms to do caching itself for performance.

I also agree. But please note we need an intermediate solution soo before we rearchitect much. I suggest we rearchitect the design once the Grid library is available. Meanwhile please have a solution but one dynamic without ifdefs and cmake build (hardwired) options.

**Comment by wdeconinck** [comment]
> > I also tend to agree with @pmaciel in fact that caching would better be done at a higher level, e.g. the Grid object, and the same could be said about e.g. an expensive Mesh generation step. That would trigger a whole design change in multiple packages.
> > e.g. MIR keeping an in-memory cache of grid objects alive, across invocations. This would then indeed also cater for any unstructured grid.
> > From a design point it is rather obscure to have to rely on plugin or specialised grid construction mechanisms to do caching itself for performance.
> 
> I also agree. But please note we need an intermediate solution soo before we rearchitect much. I suggest we rearchitect the design once the Grid library is available. Meanwhile please have a solution but one dynamic without ifdefs and cmake build (hardwired) options.

Certainly several solutions to this particular multio performance problem could co-exist.

Following the track of this atlas-orca specific approach will require an atlas-orca config file setting to enable caching (and/or environment variable).
Then, without rearchitecting designs, we can only clear the cache in the atlas-orca plugin destruction (deregistration).
This caching should preferably only be enabled for the multio-pipelines mpi-ranks; not for compute mpi-ranks.

I leave it up to @pmaciel to clarify the much preferred MIR track further with proper cache management, and how much design change this really entails.

**Comment by dsarmany** [comment]
This apporach sounds very good to me, @wdeconinck -- thank you.

**Comment by wdeconinck** [comment]
Closing as no longer needed. Thanks @MircoValentiniECMWF  for your effort though!

**Pull request closed without merging**
